idea/data-storage/src/gear/gear-event-listener.ts

```diff
@@ -0,0 +1,282 @@
+import { CreateProgramInput } from '../program/types';
+import configuration from '../config/configuration';
+import { kafkaNetworkData } from '../common/kafka-network-data';
+import { ProducerService } from '../producer/producer.service';
+import { BlockService } from '../block/block.service';
+
+const { gear } = configuration();
+
+@Injectable()
+export class GearEventListener {
+  private logger: Logger = new Logger(GearEventListener.name);
+  private api: GearApi;
+  private genesis: Hex;
+
+  constructor(
+    private programService: ProgramService,
+    private programRepository: ProgramRepo,
+    private messageService: MessageService,
+    private codeService: CodeService,
+    private codeRepository: CodeRepo,
+    private producerService: ProducerService,
+    private blockService: BlockService,
+  ) {}
+
+  public async run() {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      await this.connectGearNode();
+      const unsub = await this.listen();
+
+      await this.producerService.setKafkaNetworkData();
+
+      await new Promise((resolve) => {
+        this.api.on('error', (error) => {
+          unsub();
+          changeStatus('gearWSProvider');
+          resolve(error);
+        });
+      });
+
+      this.logger.log('⚙️ 📡 Reconnection node');
+    }
+  }
+
+  private async connectGearNode(): Promise<void> {
+    try {
+      this.api = await GearApi.create({
+        providerAddress: gear.wsProvider,
+        throwOnConnect: true,
+      });
+
+      this.genesis = this.api.genesisHash.toHex();
+
+      this.logger.log(`⚙️ Connected to ${this.api.runtimeChain} with genesis ${this.genesis}`);
+
+      kafkaNetworkData.genesis = this.genesis;
+
+      changeStatus('gearWSProvider');
+    } catch (error) {
+      console.log('api.isReady', error);
+    }
+  }
+
+  private listen() {
+    return this.api.derive.chain.subscribeNewBlocks(async (block) => {
+      const blockHash = block.createdAtHash.toHex();
+      const timestamp = (await this.api.blocks.getBlockTimestamp(block)).toNumber();
+
+      await this.handleBlocks(block, timestamp, blockHash);
+      await this.handleExtrinsics(block, timestamp);
+
+      for (const {
+        event: { data, method },
+      } of block.events) {
+        try {
+          const payload = getPayloadByGearEvent(method, data as GenericEventData);
+          if (payload !== null) {
+            await this.handleEvents(method, { ...payload, blockHash, timestamp, genesis: this.genesis });
+          }
+        } catch (error) {
+          console.error(error);
+          this.logger.warn({ method, data: data.toHuman() });
+        }
+      }
+    });
+  }
+
+  private async handleEvents(method: string, payload: any): Promise<void> {
+    const { id, genesis, timestamp } = payload;
+
+    const eventsMethod = {
+      [Keys.UserMessageSent]: async () => {
+        const createMessageDBType = plainToClass(Message, {
+          ...payload,
```

Comment:
You need to find program here by source and place it to program field of the entity